keps/sig-network/0015-20180614-SCTP-support.md

```diff
@@ -0,0 +1,291 @@
+Describe why the change is important and the benefits to users.
+The motivation section can optionally provide links to [experience reports][] to demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+--->
+### Goals
+
+Add SCTP support to Kubernetes Service and NetworkPolicy, so applications running in pods can use the native kube-dns based service discovery for SCTP based services, and their communication can be controlled via the native NetworkPolicy way.
+
+<!---
+List the specific goals of the KEP.
+How will we know that this has succeeded?
+--->
+
+### Non-Goals
+
+It is not a goal here to add SCTP support to load balancers that are provided by cloud providers. I.e. the Kubernetes user can define Services with type=LoadBalancer and Protocol=SCTP, but if the actual load balancer implementation does not support SCTP then the creation of the Service/load balancer fails.
+
+<!---
+What is out of scope for his KEP?
+Listing non-goals helps to focus discussion and make progress.
+--->
+## Proposal
+
+<!---
+This is where we get down to the nitty gritty of what the proposal actually is.
+--->
+### User Stories [optional]
+
+<!---
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of the system.
+The goal here is to make this feel real for users without getting bogged down.
+--->
+#### Service with SCTP and Virtual IP
+As a user of Kubernetes I want to define Services with Virtual IPs for my applications that use SCTP as L4 protocol on their interfaces,so client applications can use the services of my applications on top of SCTP via that Virtual IP. 
+Example:
+```
+kind: Service
+apiVersion: v1
+metadata:
+  name: my-service
+spec:
+  selector:
+    app: MyApp
+  ports:
+  - protocol: SCTP
+    port: 80
+    targetPort: 9376
+```
+
+#### Headless Service with SCTP
+As a user of Kubernetes I want to define headless Services for my applications that use SCTP as L4 protocol on their interfaces, so client applications can discover my applications in kube-dns, or via any other service discovery method that gets information about endpoints via the Kubernetes API.
+Example:
+```
+kind: Service
+apiVersion: v1
+metadata:
+  name: my-service
+spec:
+  selector:
+    app: MyApp
+  ClusterIP: "None"
+  ports:
+  - protocol: TCP
+    port: 80
+    targetPort: 9376
+```
+#### Service with SCTP without selector
+As a user of Kubernetes I want to define Services without selector for my applications that use SCTP as L4 protocol on their interfaces,so I can implement my own service controllers if I want to extend the basic functionality of Kubernetes.
+Example:
+```
+kind: Service
+apiVersion: v1
+metadata:
+  name: my-service
+spec:
+  ClusterIP: "None"
+  ports:
+  - protocol: TCP
+    port: 80
+    targetPort: 9376
+```
+#### NetworkPolicy with SCTP
```

Comment:
I think SCTP support should be pretty trivial for plugins that implement policy with iptables, OVS, or eBPF, because all the hard parts would be handled by the kernel. Like, for OpenShift, I think the complete diff would be:

    -    } else if *port.Protocol == kapi.ProtocolTCP || *port.Protocol == kapi.ProtocolUDP {
    +    } else if *port.Protocol == kapi.ProtocolTCP || *port.Protocol == kapi.ProtocolUDP || *port.Protocol == kapi.ProtocolSCTP {
             protocol = strings.ToLower(string(*port.Protocol))